ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs

```diff
@@ -1847,6 +1865,49 @@ checkLoP ConfigEnv{tracer} DynamicEnv{loPBucket} hdr kis@KnownIntersectionState{
     else do traceWith tracer $ TraceGaveLoPToken False hdr kBestBlockNo
             pure kis
 
+-- | Perform the historical rollback check; delegating to
+-- 'HistoricalRollbacks.onRollback'.
+checkHistoricalRollback ::
+     forall m blk. (IOLike m, HasHeader blk, GetHeader blk)
+  => ConfigEnv m blk
+  -> WhetherJustFoundIntersection blk
+  -> Point blk
+  -> m ()
+checkHistoricalRollback cfgEnv justFoundIntersection rollbackPoint =
+    case justFoundIntersection of
+      JustFoundIntersection intersectionPoint
+        -- Honest peers will send a rollback to the intersection point right
```

Comment:
I wonder if it'd be simpler/cleaner over all to embellish the `recvMsgIntersectFound` handlers to explicitly require/expect that `MsgRollBack` 🤔

Reply:
Yeah, that would also work, I somewhat shyed away from it as it would explicitly *require* this behavior from our server (currently, both on `main` and with the current state of this PR, we tolerate servers that don't actually send this, also see https://github.com/IntersectMBO/ouroboros-consensus/pull/1181), but I think it should be fine. Will think about it, also given that https://github.com/IntersectMBO/ouroboros-consensus/pull/1186#discussion_r1688176610 *might* supersede this.

Reply:
I implemented this [here](https://github.com/IntersectMBO/ouroboros-consensus/tree/amesgen/historical-rollbacks-alt), but it has test failures:

<details>
<summary>Details for the test failure</summary>

```
ouroboros-consensus
  ChainSyncClient
    chainSync: FAIL (0.10s)
      *** Failed! Falsified (after 1 test and 16 shrinks):
      ChainSyncClientSetup {securityParam = SecurityParam 3, clientUpdates = ClientUpdates {getClientUpdates = Schedule {getSchedule = fromList [(Tick {tickToWord64 = 1},[AddBlock (TestBlockWith {tbHash = (testHashFromList [2]), tbSlot = SlotNo 1, tbValid = Valid, tbPayload = ()}),AddBlock (TestBlockWith {tbHash = (testHashFromList [2,0]), tbSlot = SlotNo 2, tbValid = Valid, tbPayload = ()}),AddBlock (TestBlockWith {tbHash = (testHashFromList [2,0,0]), tbSlot = SlotNo 3, tbValid = Valid, tbPayload = ()}),AddBlock (TestBlockWith {tbHash = (testHashFromList [2,0,0,2]), tbSlot = SlotNo 4, tbValid = Valid, tbPayload = ()})])]}}, serverUpdates = ServerUpdates {getServerUpdates = Schedule {getSchedule = fromList [(Tick {tickToWord64 = 1},[]),(Tick {tickToWord64 = 2},[]),(Tick {tickToWord64 = 3},[AddBlock (TestBlockWith {tbHash = (testHashFromList [3]), tbSlot = SlotNo 1, tbValid = Invalid, tbPayload = ()}),SwitchFork Origin [TestBlockWith {tbHash = (testHashFromList [2]), tbSlot = SlotNo 1, tbValid = Valid, tbPayload = ()}]]),(Tick {tickToWord64 = 4},[]),(Tick {tickToWord64 = 5},[]),(Tick {tickToWord64 = 6},[]),(Tick {tickToWord64 = 7},[AddBlock (TestBlockWith {tbHash = (testHashFromList [2,3]), tbSlot = SlotNo 2, tbValid = Invalid, tbPayload = ()})]),(Tick {tickToWord64 = 8},[SwitchFork (At (Block {blockPointSlot = SlotNo 1, blockPointHash = (testHashFromList [2])})) [TestBlockWith {tbHash = (testHashFromList [2,0]), tbSlot = SlotNo 2, tbValid = Valid, tbPayload = ()}],SwitchFork (At (Block {blockPointSlot = SlotNo 1, blockPointHash = (testHashFromList [2])})) [TestBlockWith {tbHash = (testHashFromList [2,0]), tbSlot = SlotNo 2, tbValid = Valid, tbPayload = ()}],SwitchFork Origin [TestBlockWith {tbHash = (testHashFromList [0]), tbSlot = SlotNo 1, tbValid = Valid, tbPayload = ()},TestBlockWith {tbHash = (testHashFromList [0,0]), tbSlot = SlotNo 2, tbValid = Valid, tbPayload = ()},TestBlockWith {tbHash = (testHashFromList [0,0,0]), tbSlot = SlotNo 3, tbValid = Valid, tbPayload = ()}]])]}}, startTick = Tick {tickToWord64 = 3}, invalidBlocks = InvalidBlocks {getInvalidBlocks = Schedule {getSchedule = fromList [(Tick {tickToWord64 = 1},[]),(Tick {tickToWord64 = 2},[]),(Tick {tickToWord64 = 3},[]),(Tick {tickToWord64 = 4},[(testHashFromList [3]),(testHashFromList [0,2,0,0,1,0,2,0,1,3])]),(Tick {tickToWord64 = 5},[(testHashFromList [0,2,0,0,1,3]),(testHashFromList [0,2,0,0,1,0,3])]),(Tick {tickToWord64 = 6},[]),(Tick {tickToWord64 = 7},[]),(Tick {tickToWord64 = 8},[(testHashFromList [2,3])])]}}, clientSlowBy = SlotLengthTenths 0}
      ChainSyncClientSetup:
      securityParam: 3
      clientSlowBy: 0s
      --
      clockUpdates:
      1: c4|s_
      3: c_|s1
      7: c_|s2
      8: c_|s3
      --
      clientUpdates:
      1: AddBlock (H:[2],S:1,B:1), AddBlock (H:[2,0],S:2,B:2), AddBlock (H:[2,0,0],S:3,B:3), AddBlock (H:[2,0,0,2],S:4,B:4)
      --
      serverUpdates:
      3: AddBlock (H:[3],S:1,B:1), SwitchFork <- Origin -> (H:[2],S:1,B:1)
      7: AddBlock (H:[2,3],S:2,B:2)
      8: SwitchFork <- (Point 1, [2]) -> (H:[2,0],S:2,B:2), SwitchFork <- (Point 1, [2]) -> (H:[2,0],S:2,B:2), SwitchFork <- Origin -> (H:[0],S:1,B:1) (H:[0,0],S:2,B:2) (H:[0,0,0],S:3,B:3)
      --
      startTick: Tick {tickToWord64 = 3}
      invalidBlocks:
      4: [3], [0,2,0,0,1,0,2,0,1,3]
      5: [0,2,0,0,1,3], [0,2,0,0,1,0,3]
      8: [2,3]


      Client chain: Origin ] (H:[2],S:1,B:1) :> (H:[2,0],S:2,B:2) :> (H:[2,0,0],S:3,B:3) :> (H:[2,0,0,2],S:4,B:4)
      Server chain: Origin ] (H:[0],S:1,B:1) :> (H:[0,0],S:2,B:2) :> (H:[0,0,0],S:3,B:3)
      Synced fragment: (S:1; H:(testHashFromList [2])) ]
      Trace:
      (6,40s) | Protocol: Send MsgFindIntersect [At (Block {blockPointSlot = SlotNo 4, blockPointHash = (testHashFromList [2,0,0,2])}),At (Block {blockPointSlot = SlotNo 3, blockPointHash = (testHashFromList [2,0,0])}),At (Block {blockPointSlot = SlotNo 2, blockPointHash = (testHashFromList [2,0])}),At (Block {blockPointSlot = SlotNo 1, blockPointHash = (testHashFromList [2])})]
      (7,40s) | Protocol: Recv MsgIntersectFound At (Block {blockPointSlot = SlotNo 1, blockPointHash = (testHashFromList [2])}) Tip (SlotNo 1) (testHashFromList [2]) (BlockNo 1)
      (7,40s) | Client: TraceFoundIntersection (At (Block {blockPointSlot = SlotNo 1, blockPointHash = (testHashFromList [2])})) (Tip (SlotNo 4) (testHashFromList [2,0,0,2]) (BlockNo 4)) (Tip (SlotNo 1) (testHashFromList [2]) (BlockNo 1))
      (8,40s) | Protocol: Send MsgRequestNext
      (8,40s) | Protocol: Recv MsgRollBackward Origin Tip (SlotNo 3) (testHashFromList [0,0,0]) (BlockNo 3)

      Exception: NoRollbackToIntersection (At (Block {blockPointSlot = SlotNo 1, blockPointHash = (testHashFromList [2])})) (Right Origin)
      Use --quickcheck-replay="(SMGen 7391498360415002818 16079556119013927611,30)" to reproduce.
```

</details>

This is because the [reference implementation](https://github.com/IntersectMBO/ouroboros-network/blob/29899dffaaab43e82cbee381c8889563083cc2b8/ouroboros-network-protocols/testlib/Ouroboros/Network/Protocol/ChainSync/Examples.hs#L161-L167) does not adhere to the [spec](https://ouroboros-network.cardano.intersectmbo.org/pdfs/network-spec/network-spec.pdf#section.3.7):

![2024-07-24_16-22](https://github.com/user-attachments/assets/6a2f7925-7306-4aaa-9a2b-8c834885cbcc)

Should be easy to fix though.

Reply:
Superseded by using the approach from https://github.com/IntersectMBO/ouroboros-consensus/pull/1186#discussion_r1688176610